### PR TITLE
gnomeExtensions.drop-down-terminal: init at v24

### PR DESCRIPTION
**Pull request by ericdallo**
###### Motivation for this change
Add the drop down terminal GNOME shell extension.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by hedning** [comment]
This one require a few dependencies at runtime which needs fixing:

- Needs typelib from `vte`, so needs something similar to eg. this but for `vte` https://github.com/NixOS/nixpkgs/blob/373f59f4093108c814637bb39130bb245484c210/pkgs/desktops/gnome-3/core/gnome-shell-extensions/default.nix#L20-L25
- It runs stuff with `gjs` so that needs an absolute path, so similar to eg. this: https://github.com/NixOS/nixpkgs/blob/373f59f4093108c814637bb39130bb245484c210/pkgs/applications/backup/deja-dup/fix-paths.patch#L7-L8

File: pkgs/desktops/gnome-3/extensions/drop-down-terminal/default.nix [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  uuid = "drop-down-terminal@gs-extensions.zzrough.org";
+
+  installPhase = ''
+    mkdir -p $out/share/gnome-shell/extensions
+    cp -r ${uuid} $out/share/gnome-shell/extensions/
+  '';
+
+  meta = with stdenv.lib; {
+    description = "Configurable drop down terminal shell";
+    license = licenses.unfree;
```
Comment (worldofpeace) [inline pkgs/desktops/gnome-3/extensions/drop-down-terminal/default.nix]:
```suggestion
    license = licenses.gpl3;
```

https://github.com/zzrough/gs-extensions-drop-down-terminal/blob/master/COPYING ?

File: pkgs/desktops/gnome-3/extensions/drop-down-terminal/default.nix [hunk 2]
```diff
@@ -0,0 +1,27 @@
+{ stdenv, fetchFromGitHub }:
+
+stdenv.mkDerivation rec {
+  name = "gnome-shell-extension-drop-down-terminal-${version}";
```
Comment (worldofpeace) [inline pkgs/desktops/gnome-3/extensions/drop-down-terminal/default.nix]:
```suggestion
  pname = "gnome-shell-extension-drop-down-terminal";
```

when attributes `pname` and `version` are present they're automatically used to make `name`
`name = "${pname}-${version}";`

Comment (ericdallo) [reply pkgs/desktops/gnome-3/extensions/drop-down-terminal/default.nix]: Thanks, i didnt know about that!

**Comment by ericdallo** [comment]
@hedning 
I dont know how I can perform this changes. For the first one, Do I need to create a patch for my pkg and in this other pkg, somehow download and install this `typelib` from `vte`?
And for the second, I should change in the `InstalPhase` to use absolutePath right? I tried to use inside the `installPhase` `Posix.system("@coreutils@/bin/mkdir -p $out/share/gnome-shell/extensions");` but is not working :/

**Comment by hedning** [comment]
Ah, right, I only meant that the fixes needed here are the same kind of fixes that I referenced.

So the patch will look something like this:
```diff
modified   drop-down-terminal@gs-extensions.zzrough.org/extension.js
@@ -15,6 +15,8 @@
 
 // Author: Stéphane Démurget <stephane.demurget@free.fr>
 
+imports.gi.GIRepository.Repository.prepend_search_path('@vte@/lib/lib/girepository-1.0')
+
 const Lang = imports.lang;
 const Gettext = imports.gettext.domain("drop-down-terminal");
 const Mainloop = imports.mainloop;
@@ -653,7 +655,7 @@ const DropDownTerminalExtension = new Lang.Class({
         this._killingChild = false;
 
         // finds the forking arguments
-        let args = ["gjs", GLib.build_filenamev([Me.path, "terminal.js"]), Me.path];
+        let args = ["@gjs@/bin/gjs", GLib.build_filenamev([Me.path, "terminal.js"]), Me.path];
 
         // forks the process
         debug("forking '" + args.join(" ") + "'");
```

And then `@vte@` and '@gjs@` just needs to be replaced with the correct paths when building. We have a convenient utility for this. This will find occurrences of the names to the left surrounded with `@` and replace them with the expression to the right. So in nix eg. the package `vte` simply expands to its nix store path when used as a string (This can be seen in action using `nix eval nixpkgs.vte --raw`): 
```nix
patches = [
  (substituteAll {
    src = ./fix_paths.patch;
    vte = vte;
    gjs = gnome3.gjs;
  }
```

Does that make sense?

**Comment by ericdallo** [comment]
@hedning
Yep, make total sense, I'm new to this nix world, but I think I got it how this works.
I created the patch with the diff, let me know if something is missing.

One question, how do you test this and know that some libs needs absolutePath? I'm just running `nix-build -A 'gnomeExtensions.drop-down-terminal'` to check if everything its alright

File: pkgs/desktops/gnome-3/extensions/drop-down-terminal/fix_gjs.patch [hunk 3]
```diff
@@ -0,0 +1,20 @@
+--- a/drop-down-terminal@gs-extensions.zzrough.org/extension.js
++++ b/drop-down-terminal@gs-extensions.zzrough.org/extension.js
+@@ -15,6 +15,8 @@
+ 
+ // Author: Stéphane Démurget <stephane.demurget@free.fr>
+ 
++imports.gi.GIRepository.Repository.prepend_search_path('@vte@/lib/lib/girepository-1.0')
```
Comment (hedning) [inline pkgs/desktops/gnome-3/extensions/drop-down-terminal/fix_gjs.patch]: Opps, made a typo here, should only be one `lib`.

**Comment by hedning** [comment]
Right, you basically need to run things and see if they work. For regular programs that's usually as simple as `result/bin/program`.

In the case of gnome-shell extensions we need to run gnome-shell and activate the extension. Doing this in your current gnome-shell can be a bit annoying. To solve that there's two options:
- run a virtual machine
- run a nested gnome-shell session

Building a virtual machine can be very useful in general. Put something like into  `vm.nix` in your nixpkgs repo:
```nix
{ pkgs, config, ... }:
{
  environment.systemPackages = with pkgs; [
    gnomeExtensions.drop-down-terminal
  ];

  services.openssh.enable = true;
  boot.cleanTmpDir = true;

  services.xserver = {
    enable = true;

    desktopManager = {
      gnome3.enable = true;
      gnome3.debug = true;
    };

    displayManager = {
      gdm.enable = true;
      gdm.debug = true;
    };
  };

  users.users.foo = {
    isNormalUser = true;
    uid = 1000;
    password = "foobar";
    extraGroups = [ "wheel" "networkmanager" ];
    openssh.authorizedKeys.keys = [
      # put your machines public key here to get ssh access
    ];
  };

  system.stateVersion = "16.09";
}
```
And then build and run the VM like this for instance:
```bash
nixos-rebuild build-vm -I nixpkgs=./. -I nixos-config=./vm.nix
QEMU_NET_OPTS="hostfwd=tcp::2222-:22" ./result/bin/run-*-vm -m 2500
```

Then you can log in and activate the extension with gnome-tweaks.

Now that takes a while. I use this script through emacs ([session.sh](https://github.com/paperwm/gnome-shell-mode/blob/master/local/gnome-shell-mode/session.sh)) to easily launch nested gnome-shells to test and develop extensions. For that something like this should work:
```bash
nix build -f. gnomeExtensions.drop-down-terminal
session.sh wayland result/share/gnome-shell/extensions/drop-down-terminal@gs-extensions.zzrough.org/ drop-down-terminal@gs-extensions.zzrough.org
```

**Comment by ericdallo** [comment]
Uow, its a great way to test, I'll try to setup this and test this whole thing, thank you!

**Comment by ericdallo** [comment]
I fixed the `vte` path, but this time, ther is no error message, so i cannot figure out what is wrong...

When I activate the extension, only this message shows:
image

And on Looking-glass, there's no error message:
image

**Comment by hedning** [comment]
Right, looks like `terminal.js` also needs the `imports.gi.GIRepository.Repository.prepend_search_path('@vte@/lib/girepository-1.0')` preamble, as it's run in a separate process.

**Comment by ericdallo** [comment]
Finally worked!
Thank you for the incredible help and teach @hedning :)

**Comment by worldofpeace** [comment]
@ericdallo It appears you're trying to merge this into `release-19.03` and all contributions should go to `master` first. You probably want this in the current stable release, so we can backport it after it's merged to `master`.

**Comment by ericdallo** [comment]
hum, makes sense @worldofpeace , so i need to open the PR to the master, right?

**Comment by worldofpeace** [comment]
@ericdallo Likely, it's the easiest route. It also appears you're working directly on the `release-19.03` branch and not a feature one.

Best to do
```
git checkout upstream/master
git checkout -b some-branch-name
```

**Comment by ericdallo** [comment]
Got it, I'll create a branch from updated master, open a new PR with theese changes and link it here. Thank you @worldofpeace

File: pkgs/desktops/gnome-3/extensions/drop-down-terminal/default.nix [hunk 4]
```diff
@@ -15,10 +15,14 @@ stdenv.mkDerivation rec {
 
   patches = [
     (substituteAll {
-      src = ./fix_gjs.patch;
+      src = ./fix_vte_and_gjs_on_extension.patch;
       vte = vte;
       gjs = gnome3.gjs;
     })
+    (substituteAll {
+      src = ./fix_vte_on_terminal.patch;
```
Comment (hedning) [inline pkgs/desktops/gnome-3/extensions/drop-down-terminal/default.nix]: Better to just put everything into the same file tbh.

**Comment by hedning** [comment]
Right and squash everything into one commit :)

File: pkgs/desktops/gnome-3/extensions/drop-down-terminal/default.nix [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    rev = "v${version}";
+    sha256 = "1gda56xzwsa5pgmgpb7lhb3i3gqishvn84282inwvqm86afks73r";
+  };
+
+  uuid = "drop-down-terminal@gs-extensions.zzrough.org";
+
+  patches = [
+    (substituteAll {
+      src = ./fix_vte_and_gjs_on_extension.patch;
+      vte = vte;
```
Comment (worldofpeace) [inline pkgs/desktops/gnome-3/extensions/drop-down-terminal/default.nix]: You can also do
```
inherit vte;
```

which is the same as `vte = vte;`

**Comment by ericdallo** [comment]
New PR with commits squashed https://github.com/NixOS/nixpkgs/pull/70042

**Pull request closed without merging**